### PR TITLE
Refactor pointer manipulation to be independent of datatype width

### PR DESCRIPTION
**Pull request by elms**
Pointer manipulation needs to be based on the CPU native pointer type regardless of the other data type sizes. Otherwise there are compiler warnings and test failures.

Tested with `./configure CFLAGS="-DNO_64BIT" --disable-sha512
--disable-sha384 --enable-harden` on a 64-bit machine

### CODE REVIEW

**Comment by elms** [comment]
Would replace #2757

**Comment by elms** [comment]
This PR was a little larger than I first expected. I think we could further clean up some of the pointer manipulation. I'll comment on a specific place, curious what y'all think.

File: wolfcrypt/src/tfm.c [hunk 1]
```diff
@@ -1840,8 +1840,8 @@ int fp_exptmod_nb(exptModNb_t* nb, fp_int* G, fp_int* X, fp_int* P, fp_int* Y)
   #ifdef WC_NO_CACHE_RESISTANT
     err = fp_sqr(&nb->R[nb->y], &nb->R[nb->y]);
   #else
-    fp_copy((fp_int*) ( ((wolfssl_word)&nb->R[0] & wc_off_on_addr[nb->y^1]) +
-                        ((wolfssl_word)&nb->R[1] & wc_off_on_addr[nb->y]) ),
+    fp_copy((fp_int*) ( ((wc_ptr_t)&nb->R[0] & wc_off_on_addr[nb->y^1]) +
+                        ((wc_ptr_t)&nb->R[1] & wc_off_on_addr[nb->y]) ),
```
Comment (elms) [inline wolfcrypt/src/tfm.c:1844]: Is there something I'm missing or could we simplify to this and remove `wc_off_on_addr`?

```suggestion
    fp_copy((fp_int*) ( &nb->R[nb->y]),
```

File: wolfssl/wolfcrypt/types.h [hunk 2]
```diff
@@ -185,7 +193,13 @@ decouple library dependencies with standard string, memory and so on.
...
         #define MP_16BIT  /* for mp_int, mp_word needs to be twice as big as
                              mp_digit, no 64 bit type so make mp_digit 16 bit */
-        #define WC_32BIT_CPU
+#endif
+
+#ifdef HAVE_UINTPTR_T
+    #include <stdint.h>
+    typedef uintptr_t wc_ptr_t;
+#else
+    typedef size_t wc_ptr_t;
```
Comment (dgarske) [inline wolfssl/wolfcrypt/types.h:202]: Recommend a way for user to supply `wc_ptr_t`, just in case...

**Review by dgarske — CHANGES_REQUESTED** [review]
Thanks for these changes! I really like the refactor. Just a few items of feedback.

This almost fully replaces the uses of `wolfssl_word` with `wc_ptr_t`... are the ones remaining used for size not associated with a pointer?

File: wolfcrypt/src/tfm.c (hunk 1 above)
Comment (dgarske) [reply wolfcrypt/src/tfm.c:1844]: I replied... it didn't save. This is here for timing resistance and needs to remain the same.

Comment (elms) [reply wolfcrypt/src/tfm.c:1844]: interesting, I didn't include the change as I thought there was something. Any resource on why one is why one is timing resistant and the other isn't?

**Review by elms — COMMENTED** [review]
@dgarske The other uses look mostly used for alignment and padding, but also word size for xor loop `XorWords`. Not sure the impact of changing those. Maybe it's better to just change the definition of `wolfssl_word` instead of adding a new type.

**Comment by elms** [comment]
> This almost fully replaces the uses of `wolfssl_word` with `wc_ptr_t`... are the ones remaining used for size not associated with a pointer?

It's true that it replaces a majority of the uses fo `wolfssl_word`. The remaining uses are in calls to `XMEMCPY()` and `XorWords()` for buffer lengths.

**Review by dgarske — APPROVED** [review]
I have reviewed and tested this refactor. I love it. I would like @SparkiDev to also finish his review/approve then we can merge.

**Comment by SparkiDev** [comment]
Would like to see wolfssl_word usage changed in:
  wolfcrypt/test/test.c:11781
Otherwise looks good.

**Comment by SparkiDev** [comment]
Also, there are a number of size_t castings that would be better off as wc_ptr_t:
    aes.c:1738
    aes.c:3835
    md5.c:142
    md5.c:408
    memory.c:1116
    memory.c:1124
    sha256.c:501
    sha256.c:1058
    sha.c:257
    sha.c:615

I'll fix the SP code when this is merged.

**Comment by elms** [comment]
@SparkiDev updated all the ones you pointed out. Thanks for pointing those out.

**Review by SparkiDev — APPROVED** [review]

**Review by dgarske — APPROVED** [review]

**Pull request merged**